numba/array_analysis.py

```diff
@@ -0,0 +1,718 @@
+from __future__ import print_function, division, absolute_import
+import types as pytypes # avoid confusion with numba.types
+from numba import ir
+from numba.ir_utils import *
+#from numba.annotations import type_annotations
+from numba import types, config
+from numba.types.npytypes import *
+from numba.typing import npydecl
+import collections
+import copy
+
+import numpy
+MAP_TYPES = [numpy.ufunc]
+
+class ArrayAnalysis(object):
+    """Analyzes Numpy array computations for properties such as shapes
+    and equivalence classes.
+    """
+
+    def __init__(self, func_ir, typemap, calltypes):
+        self.func_ir = func_ir
+        self.typemap = typemap
+        self.calltypes = calltypes
+        self.next_eq_class = 1
+        # equivalence classes for each dimension of each array are saved
+        # string to tuple of class numbers
+        # example: {'A':[1,2]}
+        #          {1:[n,a],2:[k,m,3]}
```

Comment:
I added more comments for clarification. Let's discuss it sometime if it's not clear yet.